File: allcode/2301-2400/2327peopleAwareOfSecret.py
# 在第 1 天，有一个人发现了一个秘密。
#
# 给你一个整数 delay ，表示每个人会在发现秘密后的 delay 天之后，每天 给一个新的人 分享 秘密。同时给你一个整数 forget ，表示每个人在发现秘密 forget 天之后会 忘记 这个秘密。一个人 不能 在忘记秘密那一天及之后的日子里分享秘密。
#
# 给你一个整数 n ，请你返回在第 n 天结束时，知道秘密的人数。由于答案可能会很大，请你将结果对 109 + 7 取余 后返回。
#
#  
#
# 示例 1：
#
# 输入：n = 6, delay = 2, forget = 4
# 输出：5
# 解释：
# 第 1 天：假设第一个人叫 A 。（一个人知道秘密）
# 第 2 天：A 是唯一一个知道秘密的人。（一个人知道秘密）
# 第 3 天：A 把秘密分享给 B 。（两个人知道秘密）
# 第 4 天：A 把秘密分享给一个新的人 C 。（三个人知道秘密）
# 第 5 天：A 忘记了秘密，B 把秘密分享给一个新的人 D 。（三个人知道秘密）
# 第 6 天：B 把秘密分享给 E，C 把秘密分享给 F 。（五个人知道秘密）
# 示例 2：
#
# 输入：n = 4, delay = 1, forget = 3
# 输出：6
# 解释：
# 第 1 天：第一个知道秘密的人为 A 。（一个人知道秘密）
# 第 2 天：A 把秘密分享给 B 。（两个人知道秘密）
# 第 3 天：A 和 B 把秘密分享给 2 个新的人 C 和 D 。（四个人知道秘密）
# 第 4 天：A 忘记了秘密，B、C、D 分别分享给 3 个新的人。（六个人知道秘密）
#  
#
# 提示：
#
# 2 <= n <= 1000
# 1 <= delay < forget <= n

from leetcode.allcode.competition.mypackage import *

class Solution:
    def peopleAwareOfSecret(self, n: int, delay: int, forget: int) -> int:
        N = int(1e9 + 7)
        f = [0] * n
        g = [0] * n
        for i in range(delay):
            f[i] = 1
        g[0] = 1
        for i in range(delay, forget):
            g[i] = sum(g[:i - delay + 1]) % N
            f[i] = (g[i] + f[i - 1]) % N
        for i in range(forget, n):
            s = 0
            for j in range(i - forget + 1, i - delay + 1):
                s += g[j]
                s %= N
            g[i] = s
            f[i] = (g[i] - g[i - forget] + f[i - 1]) % N
        print(g)
        print(f)
        return f[-1]


so = Solution()
print(so.peopleAwareOfSecret(n = 7, delay = 3, forget = 5))
print(so.peopleAwareOfSecret(n = 6, delay = 2, forget = 4))
print(so.peopleAwareOfSecret(n = 4, delay = 1, forget = 3))

File: c++/src/2000-2999/2327peopleAwareOfSecret.cpp
#include "lc_pub.h"


class Solution {
    public:
    int peopleAwareOfSecret(int n, int delay, int forget) {
        int MOD=1'000'000'007;
        vector<long long> diff(n, 0);   // 知道秘密的人数的差分数组，实际只存储需要减少的部分
        vector<long long> diff2(n, 0);  // 可以传递秘密的人数的差分数组
        long long num1=1, num2=0;
        diff2[delay] = 1;
        if (forget < n) {
            diff[forget] -= 1;
            diff2[forget] -= 1;
        }
        for (int i=0;i<n;i++) {
            num2=(num2+diff2[i]+MOD)%MOD;  // 当天可以传递秘密的人数
            num1=(num1+diff[i]+num2+MOD)%MOD;  // 当天知道秘密的人数增加了 diff[i]+num2，其中nums2是增加的部分，diff[i]是通过差分需要减的部分
            if (i+delay<n) {
                diff2[i+delay]+=num2;
            }
            if (i+forget<n) {
                diff[i+forget]-=num2;
                diff2[i+forget]-=num2;
            }
            // cout<<num1<<","<<num2<<endl;
        }
        return num1;
    }
    };
    
int main()
{
    cout<<"test let us start! %s" << __cplusplus <<std::endl;

    Solution so;
    cout << so.peopleAwareOfSecret(6, 2, 4) << endl;
    return 0;
}
